We used site allele frequency files to calculate Fst and dxy between the CH and CO source populations (CHST, COSK) and to estimate nucleotide diversity (π) and Tajima’s D over 50kb non-overlapping windows throughout the 1.9G T. conura genome. The two-dimensional folded site frequency spectrum (SFS) was used to calculate Fst using the Bhatia estimator (ANGSD v. 0.940, realSFS fst index; realSFS fst stats2; Korneliussen et al., 2014). To calculate dxy, we first re-calculated the 2DSFS for every 50kb window, then calculated dxy using a modified version of dxy_wsfs.py script from (D. Marques, https://github.com/marqueda/PopGenCode/blob/master/dxy_wsfs.py, accessed Nov. 2023) modified to run in R (dxy_wsfs.R; Supplementary Materials). We also used ANGSD to calculate π and Tajima’s D over 50kb windows.
